File: src/diffuseur.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <fcntl.h>
#include <pthread.h>
#include <errno.h>
#include <sys/types.h>
#include <netdb.h>
#include "diffuseur.h"
#include "netradio.h"

#define min(a, b) (a <= b ? a : b)

// Les informations du diffuseur actuel
diff_info di;

// La liste de message
char *msgList[MAX_MSG];
// Les index nécessaire
int NUM_MSG, MSG_INDEX, NBR_SENT = 0;
// Le verrou
pthread_mutex_t msgLock = PTHREAD_MUTEX_INITIALIZER;

int start();
int add_message(char *msg);
void mess(int sock);
void last(int sock);
void print_infos(diff_info di);

int main(int argc, char *argv[])
{   
    if (argc < 2)
    {
        printf("Veuillez indiquer un fichier de settings.\n");
        return -1;
    }
    if(argc < 3){
        printf("Veuillez indiquer un fichier de messages.\n");
        return -1;
    }

    // On récupère les réglages
    di = load_settings(argv[1]);
    di.ip2 = get_host_address();

    print_infos(di);

    // Puis les messages de base
    int size = 0;
    char **msgs = get_msgs(argv[2], &size);

    // Puis on charge les messages
    for (int i = 0; i < size; i++)
    {
        char buff[ID + 1 + strlen(msgs[i]) + 1];
        memset(buff, 0, ID + 1 + strlen(msgs[i]) + 1);
        strncpy(buff, di.id, ID);
        buff[ID] = ' ';

        memcpy(buff + ID + 1, msgs[i], strlen(msgs[i]));

        add_message(buff);
    }

    // Et on lance le diffuseur
    start();
    return 0;
}

void incr_msg()
{
    NUM_MSG = (NUM_MSG + 1) % MAX_MSG;
}

void incr_msg_index()
{
    MSG_INDEX = (MSG_INDEX + 1) % NUM_MSG;
}

int add_message(char *msg)
{
    pthread_mutex_lock(&msgLock);
    msgList[NUM_MSG] = fill_with_sharp(msg, ID + 1 + MESS);
    incr_msg();
    pthread_mutex_unlock(&msgLock);

    return 0;
}

void *diffuse()
{
    int sock = socket(PF_INET, SOCK_DGRAM, 0);
    struct addrinfo *fi;
    struct addrinfo h;
    memset(&h, 0, sizeof(struct addrinfo));
    h.ai_family = AF_INET;
    h.ai_socktype = SOCK_DGRAM;
    char port[5];
    memset(port, 0, 5);
    sprintf(port, "%d", di.port1);
    int r = getaddrinfo(di.ipmulti, port, NULL, &fi);
    if (r != 0 || fi == NULL)
    {
        perror("getaddrinfo");
        exit(-1);
    }
    struct sockaddr *saddr = fi->ai_addr;
    while (1)
    {
        pthread_mutex_lock(&msgLock);
        if (NUM_MSG == 0)
        {
            pthread_mutex_unlock(&msgLock);
            continue;
        }
        char *msg = msgList[MSG_INDEX];
        incr_msg_index();
        pthread_mutex_unlock(&msgLock);

        int size = 4 + 1 + NUMMESS + 1 + ID + 1 + MESS + 2;
        char buff[size];
        memset(buff, 0, size);
        char *num = fill_with_zeros(NBR_SENT, NUMMESS);
        sprintf(buff, "DIFF %s %s\r\n", num, msg);

        sendto(sock, buff, size, 0, saddr, (socklen_t)sizeof(struct sockaddr_in));
        NBR_SENT = (NBR_SENT + 1) % MAX_MSG;

        free(num);
        sleep(SLEEP_TIME);
    }
    return NULL;
}

void *client_routine(void *adr)
{
    int sock = *((int *)adr);

    char cmd[5];
    memset(cmd, 0, 5);

    recv(sock, cmd, 4 * sizeof(char), 0);

    if (!strcmp(cmd, "LAST"))
    {
        last(sock);
    }
    else if (!strcmp(cmd, "MESS"))
    {
        mess(sock);
    }
    else
    {
        close(sock);
    }
    return NULL;
}

void mess(int sock)
{
    // On saute l'espace
    char space;
    recv(sock, &space, sizeof(char), 0);
    int size = ID + 1 + MESS + 2;
    char buff[size];
    memset(buff, 0, size);

    if (recv(sock, buff, size, 0) != size)
    {
        perror("recv");
        close(sock);
        exit(-1);
    }

    buff[size - 1] = 0;
    buff[size - 2] = 0;
    add_message(buff);

    send(sock, "ACKM\r\n", sizeof(char) * 6, 0);

    close(sock);
}

void last(int sock)
{
    char space;
    recv(sock, &space, sizeof(char), 0);

    int size = NBMESS + 2;
    char nb[size];
    memset(nb, 0, size);

    if (recv(sock, nb, size, 0) != size)
    {
        perror("recv");
        close(sock);
        exit(-1);
    }

    pthread_mutex_lock(&msgLock);
    int n = min(atoi(nb), NBR_SENT + 1);

    for (int i = 0; i < n; i++)
    {
        char *s = msgList[(NBR_SENT - i) % NUM_MSG];
        int size = 4 + 1 + NUMMESS + 1 + ID + 1 + MESS + 2;

        char oldm[size];
        memset(oldm, 0, size);
        sprintf(oldm, "OLDM %s %s\r\n", fill_with_zeros(NBR_SENT - i, NUMMESS), s);

        send(sock, oldm, size, 0);
    }
    pthread_mutex_unlock(&msgLock);

    send(sock, "ENDM\r\n", 6, 0);

    close(sock);
}

void *receive()
{
    // On créer un server
    int sock = create_tcp_server(di.port2);

    // On écoute sur la socket
    if (listen(sock, 0))
    {
        perror("listen");
        exit(-1);
    }

    // Puis on récupère toutes les connexions entrantes
    while (1)
    {
        struct sockaddr_in caller;
        socklen_t size = sizeof(caller);

        int sock_cli = accept(sock, (struct sockaddr *)&caller, &size);
        if (sock_cli < 0)
            continue;

        pthread_t th_client;
        if (pthread_create(&th_client, NULL, client_routine, &sock_cli) != 0)
        {
            perror("pthread_create diffuse");
            exit(-1);
        }
    }
    return NULL;
}

int start()
{
    pthread_t th_diff;
    if (pthread_create(&th_diff, NULL, diffuse, NULL) != 0)
    {
        perror("pthread_create diffuse");
        exit(-1);
    }
    pthread_t th_receive;
    if (pthread_create(&th_receive, NULL, receive, NULL) != 0)
    {
        perror("pthread_create receive");
        exit(-1);
    }
    read(STDIN_FILENO, NULL, 1);
    return 0;
}

void print_infos(diff_info di){
    printf("#-----------------------------------#\n");
    printf(" * id : %s\n", di.id);
    printf(" * ip multi diffusion : %s\n",di.ipmulti);
    printf(" * port multi diffusion : %d\n",di.port1);
    printf(" * ip TCP : %s\n",di.ip2);
    printf(" * port TCP : %d\n", di.port2);
    printf("#-----------------------------------#\n");
}

File: src/Java/Diffuseur.java
import java.io.*;
import java.net.*;
import java.util.HashMap;
import java.util.Scanner;
import java.util.concurrent.atomic.AtomicInteger;

public class Diffuseur {
    private final String id;
    private String[] msgList;
    private final int port1, port2;
    private final String ipmulti, ip2;

    private AtomicInteger NUM_MSG = new AtomicInteger(0);
    private AtomicInteger MSG_INDEX = new AtomicInteger(0);
    private int NBR_SENT = 0;

    private static int MAX_MSG = 10000;
    private static int SLEEP_TIME = 1000;

    // Diffuseur identifié par :
    // un id, un port et une adresse multi diff, et un port pour la communication
    // connectée
    public Diffuseur(String id, int port1, int port2, String ipmulti) throws IOException {
        this.id = NetRadio.fillWithSharp(id, NetRadio.ID);
        this.port1 = port1;
        this.port2 = port2;
        this.ipmulti = ipmulti;
        this.ip2 = InetAddress.getLocalHost().getHostAddress();
        msgList = new String[MAX_MSG];

        startMessage(id, ipmulti, ip2, port1, port2);
    }

    public void loadMessage(String filename) throws IOException {
        String[] msgs = FileLoader.loadMessages(filename);
        synchronized (msgList) {
            for (int i = 0; i < Math.min(msgs.length, (MAX_MSG - MSG_INDEX.get())); i++) {
                addMessage(id + " " + msgs[i]);
            }
        }
    }

    private void addMessage(String msg) {
        synchronized (msgList) {
            msgList[NUM_MSG.get()] = NetRadio.fillWithSharp(msg, NetRadio.ID + 1 + NetRadio.MESS);
            incrMsg();
        }
    }

    // Permet d'incrémenter le nombre de message
    private void incrMsg() {
        NUM_MSG.set((NUM_MSG.incrementAndGet()) % 10000);
    }

    private void incrMsgIndex() {
        MSG_INDEX.set((MSG_INDEX.incrementAndGet()) % NUM_MSG.get());
    }

    // Met s dans le tableau data
    // - Remplit de \0 si trop petit
    // - Tronque si trop grand
    private void stringInBytes(byte[] data, String s) {
        byte[] newS = s.getBytes();
        for (int i = 0; i < data.length; i++) {
            if (i < newS.length)
                data[i] = newS[i];
            else
                data[i] = '\0';
        }
    }

    private String getMsg() {
        synchronized (msgList) {
            return msgList[MSG_INDEX.get()];
        }
    }

    public void start() throws UnknownHostException, IOException {
        // On créer un thread qui va diffuser
        new Thread(() -> diffuse()).start();
        // Puis on créer un nouveau thread pour récuperer les connexions TCP
        new Thread(() -> {
            try (ServerSocket sso = new ServerSocket(port2)) {
                while (true) {
                    Socket sock = sso.accept();
                    // On créer un nouveau thread pour la connexion entrante
                    new Thread(() -> {
                        try (BufferedReader br = new BufferedReader(new InputStreamReader(sock.getInputStream()))) {
                            char[] command = new char[4];
                            br.read(command, 0, 4);
                            // On regarde quelle commande est effectuée
                            switch (String.valueOf(command)) {
                            case "LAST":
                                lastMsgs(sock, br);
                                break;
                            case "MESS":
                                addMessageToList(sock, br);
                                break;
                            default:
                                sock.close();
                                break;
                            }
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }).start();
                }
            } catch (Exception e) {
                System.out.println("An error as occured : " + e.getMessage());
            }
        }).start();
        // Puis on lit en boucle les inputs pour récuperer les demandes de REGI
        Scanner s = new Scanner(System.in);
        while (true) {
            String rq = s.nextLine();
            String[] opts = rq.split(" ");
            if (opts[0].equals("REGI")) {
                new Thread(() -> {
                    try {
                        String ip = opts[1];
                        int port = Integer.valueOf(opts[2]);
                        Socket sock = new Socket(ip, port);
                        PrintWriter pw = new PrintWriter(new OutputStreamWriter(sock.getOutputStream()));
                        pw.print("REGI " + this.id + " " + NetRadio.normalizeIp(this.ipmulti) + " " + this.port1 + " "
                                + NetRadio.normalizeIp(this.ip2) + " " + this.port2 + "\r\n");
                        pw.flush();
                        BufferedReader br = new BufferedReader(new InputStreamReader(sock.getInputStream()));
                        char[] data = new char[6];
                        br.read(data, 0, 6);
                        if (String.valueOf(data).equals("REOK\r\n")) {
                            System.out.println("Enregistrement réussi !");
                            workWithGestionnaire(sock);
                        } else {
                            System.out.println("Echec de l'enregistrement...");
                            sock.close();
                        }
                    } catch (Exception e) {
                        System.out.println("Erreur lors de la reception de la commande...");
                    }
                }).start();
            }
        }
    }

    private void workWithGestionnaire(Socket socket) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        PrintWriter pw = new PrintWriter(new OutputStreamWriter(socket.getOutputStream()));
        while (true) {
            char[] data = new char[6];
            br.read(data, 0, 6);
            if (String.valueOf(data).equals("RUOK\r\n")) {
                pw.print("IMOK\r\n");
                pw.flush();
            } else {
                break;
            }
        }
        socket.close();
    }

    // Fonction utilisée pour la diffusion
    private void diffuse() {
        try (DatagramSocket dso = new DatagramSocket()) {
            byte[] data = new byte[4 + 1 + NetRadio.NUMMESS + 1 + NetRadio.ID + 1 + NetRadio.MESS + 2];
            while (true) {
                // On envoie un message en multi-diffusion
                InetSocketAddress ia = new InetSocketAddress(ipmulti, port1);

                String msg = getMsg();
                stringInBytes(data, "DIFF " + NetRadio.fillWithZero(NBR_SENT, NetRadio.NUMMESS) + " " + msg + "\r\n");
                incrMsgIndex();
                NBR_SENT = (NBR_SENT + 1) % MAX_MSG;

                DatagramPacket paquet = new DatagramPacket(data, data.length, ia);
                dso.send(paquet);

                Thread.sleep(SLEEP_TIME);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public synchronized void lastMsgs(Socket sock, BufferedReader br) throws IOException {
        // On saute l'espace
        br.read();
        char[] nb = new char[5];
        // On récupère le nombre de message demandé
        if (br.read(nb, 0, 5) < 5) {
            br.close();
            return;
        }

        PrintWriter pw = new PrintWriter(new OutputStreamWriter(sock.getOutputStream()));
        synchronized (msgList) {
            // On prend une copie de NBR_SENT, qui ne bougera pas
            // (une sorte d'image figée au moment de la requête)
            int nb_sent_cpy = NBR_SENT;
            // Pour ne pas dépasser la taille du nombre de message diffusé
            int nbres = Math.min(Integer.valueOf(String.valueOf(nb).strip()), nb_sent_cpy + 1);
            // Puis on envoie les `nbres` derniers messages
            for (int i = 0; i < nbres; i++) {
                String s = "OLDM " + NetRadio.fillWithZero(nb_sent_cpy - i, NetRadio.NUMMESS) + " "
                        + msgList[(nb_sent_cpy - i) % NUM_MSG.get()] + "\r\n";
                byte[] data = new byte[4 + 1 + NetRadio.NUMMESS + 1 + NetRadio.ID + 1 + NetRadio.MESS + 2];
                stringInBytes(data, s);
                // Puis on envoie les données
                pw.print(new String(data));
                pw.flush();
            }
        }
        // On finit par envoyer `ENDM` et on ferme la connexion
        pw.print("ENDM\r\n");
        pw.flush();
        sock.close();
    }

    void addMessageToList(Socket sock, BufferedReader br) throws IOException {
        // On saute l'espace
        br.read();

        char[] msgChars = new char[NetRadio.ID + 1 + NetRadio.MESS + 2];
        br.read(msgChars, 0, NetRadio.ID + 1 + NetRadio.MESS + 2);
        addMessage(String.valueOf(msgChars).strip());

        PrintWriter pw = new PrintWriter(new OutputStreamWriter(sock.getOutputStream()));
        pw.print("ACKM\r\n");
        pw.flush();

        sock.close();
    }

    private static void startMessage(String id, String ip1, String ip2, int port1, int port2) {
        System.out.println("#-----------------------------------#");
        System.out.println(" * id : " + id);
        System.out.println(" * ip multi diffusion : " + ip1);
        System.out.println(" * port multi diffusion : " + port1);
        System.out.println(" * ip TCP : " + ip2);
        System.out.println(" * port TCP : " + port2);
        System.out.println("#-----------------------------------#");
    }

    public static void main(String[] args) throws IOException {
        if (args.length < 1) {
            System.err.println("Veuillez indiquer un fichier de réglage");
            return;
        } else if (args.length < 2) {
            System.err.println("Veuillez indiquer un fichier de message");
            return;
        }
        HashMap<String, String> settings = (HashMap<String, String>) FileLoader.loadSettings(args[0]);

        String id = settings.get("id");
        String ip1 = settings.get("ip1");
        int port1 = Integer.valueOf(settings.get("port1"));
        int port2 = Integer.valueOf(settings.get("port2"));
        Diffuseur diff = new Diffuseur(id, port1, port2, ip1);
        diff.loadMessage(args[1]);
        diff.start();
    }
}